src/index.ts

```diff
@@ -30,6 +31,15 @@ export type Match<Record, Union> = {
   <A>(variant: Union, cases: MatchCases<Record, Union, A>): A;
 };
 
+export type UpdateCases<Record> = Partial<
+  { [T in keyof Record]: (value: Record[T]) => Partial<Record[T]> }
```

Comment:
Using `Partial` for the return type here is unsafe because it allows overwriting values with `undefined`. I'm not sure if inference would be able to handle using `Pick` though... Perhaps we should statically disallow values to be `undefined`, which we're already doing in the case of `SingleValueVariants`.

Reply:
I think it is possible to not use Partial + remove this magic check for 'isObject'.
I think the way to do it is this:
```typescript
function update<T>(val:T):T | Part<T> {}; // Not Partial
type Part<T> = {_opaque:T}//  is opaque for user.

// helper function to create a part
function part<K extends keyof T, T>(p: Pick<K,T>) : Part<T> {
   // tag that this is a part object for implementation
   return  {__part: p} as any as Part<T>
}

// simplified implementation for update
function updateObject(oldV, newV) {
if (typeof newV === 'object'  && newV.__part)  {
   // merge objects
   return merged.
   }
   // if it is not a part it has to be a full object.
   return newV;
} 
```
So you either have to provide a complete object (not Partial<T>) or use a helper function.

```typescript
const Data = unionize({AB: ofType<{ a: number; b: string }>()});
const ab = Data.AB({ a: 1, b: 'hi' });
Data.update(ab, { AB: ({ b }) => part({ b: b + '!' }) }); //note usage of 'part'
Data.update(ab, { AB: ({ a, b }) => ({a, b: b + '!' }) }); // or provide a full shape
```
Not 100% that it is worth complexity. But thinking about updates again, I would either prefer a solution with 'part' function without 'isObject' magic or forbid partial updates. Note that in that case it is still useful:
* don't have to provide default
* don't have to call type constructor
```typescript
// a bit verbose
Data.match(ab, { AB: ({ a, b }) => Data.AB({a, b: b + '!' }), default: i=>i })
```